### ISSUE TITLE
Functor instance breaks laws

### ISSUE THREAD

**Issue by paluh** [d86leader/purescript-undefined-or#4]
Hi @d86leader!

Thanks for putting lib together.

I think that `Functor`doesn't hold (and from that all other instances):

* `UndefinedOr undefinedVal` on the representational level is just `undefined`.
* `map` just wraps a function result in a `UndefinedOr` or returns `undefinedVal` directly.
* So `map (const undefinedVal)` is`undefinedVal` for both cases.
* So `map (const 8) <<< map (const undefinedVal)` all the time results in `undefinedVal`.
* But `map (const 8 <<< const undefinedVal)` can result in something when applied to non `undefined` value because `map (const 8 <<< const undefinedVal) == map (const 8)`

In other words:

```
map (const 8 <<< const undefined) (toUndefined "a") == toUndefined 8
```

but

```
map (const 8) <<< map (const undefinedVal) (toUndefined "a") == undefinedVal
```

If the above is true then this `Functor` breaks the law:

```
map f <<< map g == map (f <<< g)
```

**Comment by maurges** [comment]
Good catch, that was really sloppy programming on my part. Doesn't seem hard to fix, I just need to get the purescript toolchain again. I wonder how is purescript-sans-npm is going nowadays

**Comment by paluh** [comment]
Unfortunately I think that you can fix it only by dropping instances ;-)

**Comment by paluh** [comment]
By the way - this is not my catch - it is pretty known fact about `Maybe a` vs `Nullable a`. `null` and `undefined` are isomorphic so the same should hold for `UndefinedOr a` I think.

**Comment by maurges** [comment]
Ohh, I see what you're mean, the fact that `Just Nothing` and `Nothning` are both homomorphic to `undefined`. God damn it.

**Comment by maurges** [comment]
I don't really want to drop the instances though, it makes programming this more convenient. But hm, the most useful one I think is the Alternative, and I can just provide a custom operator for that instead. The biggest problem with making this decision is that my investment in this matter is pretty low now, as I haven't programmed in purescript in a long time.

**Comment by paluh** [comment]
Please note that broken instances can affect people in really surprising way and at different level of their application. There are higher order composition constructs `Compose` for `Functor` and `Applicative` which gonna be broken for every stack (like composition of `UndefinedOr` with `Effect`, `State` or whatever) which includes your type.
In other words I would drop them or put clear warning in the `README`.

**Comment by maurges** [comment]
Well, I do already make a disclaimer in the readme, and specifically disallow monad instance to deter people from using this in too advanced ways, and the functor+alternative instances are just for convenient overloads, not the advanced usage.
Ok, the [readme for functor](https://pursuit.purescript.org/packages/purescript-prelude/6.0.0/docs/Data.Functor#t:Functor) says that it *must* satisfy the the equations. So I better do it as to not break the law.

**Comment by CarstenKoenig** [comment]
I can see how the functions themselves can be useful so maybe just have a `map`, `apply`, `pure`, `alt` function together with an `empty` value exported from the module and add a disclaimer in the doc-comments to that functions? It will not break any laws and probably most use-cases will just have to add a `UndefinedOr.` in front of functions.

Yeah it would stop working with other higher-order functions but those are most likely assuming *lawfulness*.

If that is ok with you there is no need for you to install the toolchain again - I'd volunteer to do it for you.

**Comment by maurges** [comment]
Yeah, makes sense, I wanted to do something like that. If you were to do it, I would gladly merge it

**Comment by maurges** [comment]
Removed instances in #7
